### PR TITLE
Add task-based history export tracking

### LINKED ISSUE #14511: Switch history export (UI?) to use STS
I don't think history exports were meant to be stable, there's certainly no reasonable management behind it, other than removing them regularly with the cleanup scripts. Therefore we can use our implementation of Short-Term Storage (STS) to allow the export of a history archive for as long as its STS request id exists.

xref #14459

### PR DESCRIPTION
**Pull request by davelopez**
Requires #15090 (To pass the remaining integration test failure)
Fixes #14511
Related to #14771

## TLDR

This PR enables the tracking of every history export (and potentially any other exportable object) in the database by adding a new table containing structured metadata about the export context.
It also creates a new `HistoryExport` component in the UI that is based on the asynchronous task (Celery) system instead of the regular job system for handling exports and short-term downloads (#14511).

## Backend Changes

- Adds a new `store_export_association` table to track any object export associated with a task. This table has an `export_metadata` column containing information about the export context (request and result) in a structured JSON (modeled with Pydantic objects).

- Adds a new `StoreExportTracker` class to associate export metadata with any exportable object.

- Refactors the `HistoryExportView` to `HistoryExportManager` for managing the exports associated with histories.

- Adapts the `model_stores` to generate and associate the export metadata to history downloads and remote exports.

- Refactors the Tasks API to use an abstract `AsyncTaskManager` and the corresponding `CeleryAsyncTaskManager` implementation to query tasks status.

- Modifies the `/api/histories/{id}/exports` endpoint to accept a new type `application/vnd.galaxy.task.export+json` that will return the new task-based export associations. It will continue returning the previous "JEHAs" using the default request accept type so it's backward compatible.

- Some other small refactorings for the schemas, removing unused code, etc.

## Frontend Changes

Added several new components to display information on the last export and previous ones along with actions to download them again or re-import them from remote sources, or even choose the preferred archive format. They provide similar functionality as the previous components (that still are used when Celery is not enabled) and include a bit more related to the export tracking.

### Choose Export Options

Now you can decide which export format and dataset files want to include in the archive.

image

### Direct Download (Temporal)

When you use the direct download option, it will be reused (the package won't be regenerated) whenever you try to download it again until:

- The download link expires
- The history changes
- The export parameter change

  image

### Export to Remote File Source (Long term)

When you export your history to a remote file source a reference will be stored in the record and you can reimport it later as long as the remote file source maintains the exported package.
image

### List of previous exports

You can download or reimport copies (snapshots) of previous versions of the history as long as they have not expired or the remote source maintains them.
image

## TODO

- [x] Display options to select/change the export format
- [x] Add a bunch of tests
- [ ] Make the UI reactive to history changes (requires history store (Pinia) for composition API)

## How to test the changes?

(Select all options that apply)

- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License

- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/schema/schema.py [hunk 1]
```diff
@@ -1410,8 +1414,40 @@ class JobExportHistoryArchiveModel(Model):
...
+
+
+AnyExportHistoryArchiveModel = Annotated[
+    Union[JobExportHistoryArchiveModel, TaskExportHistoryArchiveModel], Field(discriminator="type")
+]
+
+
 class JobExportHistoryArchiveCollection(Model):
-    __root__: List[JobExportHistoryArchiveModel]
+    __root__: List[AnyExportHistoryArchiveModel]
```
Comment (davelopez) [inline lib/galaxy/schema/schema.py]: I'm trying to decide if returning both types of "views" (Job or Task) of a history export will be too much of a breaking change for the API...

Will it be more desirable to create a dedicated route for the task-based exports (like `api/histories/{id}/exports/tasks`) and leave the current `api/histories/{id}/exports` just for the job-based one so the API doesn't change that much?
OTOH I feel like it will be a bit harder to maintain in the client later if you need to request both endpoints to know how the history was exported... :thinking:

Comment (davelopez) [reply lib/galaxy/schema/schema.py]: I guess the question boils down to... what is the best approach to associate export metadata to an exported "object" (not necessarily only histories) so we can track when this object was exported and where (for recovery)?

I was hoping to use `api/histories/{id}/exports` for the particular case of histories because it makes a lot of sense semantically. But maybe we should use a more general API independent of the object type?

Comment (davelopez) [reply lib/galaxy/schema/schema.py]: Following up with the discussion yesterday in the backend meeting about this...

1 - Is it a better idea to create a new `GET /api/histories/{id}/store_exports` endpoint instead of trying to reuse `GET /api/histories/{id}/exports`? this **store_exports** endpoint will have an equivalent for all objects that support the **write_store** and **from_store** endpoints (like invocations)

2 - Assuming we want to use case (1). How do we want to persist the export metadata for all these objects (histories, invocations, etc)? Creating a new "store_export_association" table that tracks all the exports (excluding simple downloads) for all types of exported objects? 

3 - Do we want to deprecate the job-based export for histories and use only the task-based one at some point? Meaning we should update the current history export UI to align it more with #14606?

Comment (mvdbeek) [reply lib/galaxy/schema/schema.py]: > Creating a new "store_export_association" table that tracks all the exports (excluding simple downloads) for all types of exported objects?

I think that's a solid yes.

> this **store_exports** endpoint will have an equivalent for all objects that support the **write_store** and **from_store** endpoints (like invocations)

You mean we'll also have `/api/invocations/{id}/store_exports` ? This sounds good to me, my one concern is that `store` feels like an implementation detail consumers wouldn't need to know about. Is there any chance we can gloss over the differences and keep a backwards compatible API at `/api/histories/{id}/exports` ?

> Do we want to deprecate the job-based export for histories and use only the task-based one at some point? Meaning we should update the current history export UI to align it more with #14606?

Yes IMO. That would close https://github.com/galaxyproject/galaxy/issues/14511

Comment (davelopez) [reply lib/galaxy/schema/schema.py]: Thank you @mvdbeek, that helps a lot!

> You mean we'll also have /api/invocations/{id}/store_exports

Yes, that is what I meant :)

> Is there any chance we can gloss over the differences and keep a backwards compatible API at /api/histories/{id}/exports

The only solution that pops into my mind is using a custom content-type for returning just the task-based exports and fallback to the job-based one with the default content-type. Will that be OK in this case?

> Yes IMO. That would close https://github.com/galaxyproject/galaxy/issues/14511

Oh! great thanks! I lost track of this one :)

Comment (davelopez) [reply lib/galaxy/schema/schema.py]: I'm happy now using a different accept content type to make the `/api/histories/{id}/exports` compatible with both export modes, it feels cleaner and works great with the interactive API docs :)

**Comment by davelopez** [comment]
This one should be ready for review now :smiley: 

I will keep it in draft mode because the integration test `TestHistoryImportExportFtpSeleniumIntegrationWithTasks::test_history_export_tracking` will fail until https://github.com/galaxyproject/galaxy/pull/15090 is merged.

File: lib/galaxy_test/selenium/test_history_export.py [hunk 2]
```diff
@@ -7,6 +7,10 @@
 class TestHistoryExport(SeleniumTestCase):
     ensure_registered = True
 
+    @classmethod
+    def handle_galaxy_config_kwds(cls, config):
+        config["enable_celery_tasks"] = False
```
Comment (jmchilton) [inline lib/galaxy_test/selenium/test_history_export.py:12]: Non-integration tests shouldn't override this. Previously I don't think they even had this method. Is this definitely being used?

File: lib/galaxy/webapps/galaxy/api/histories.py [hunk 3]
```diff
@@ -379,19 +412,23 @@ def archive_export(
...
+        history_id: DecodedDatabaseIdField = HistoryIDPathParam,
         jeha_id: Union[DecodedDatabaseIdField, LatestLiteral] = JehaIDPathParam,
     ):
         """
         See ``PUT /api/histories/{id}/exports`` to initiate the creation
         of the history export - when ready, that route will return 200 status
         code (instead of 202) and this route can be used to download the archive.
+
+        **Deprecation notice**: Please use `/api/histories/{id}/prepare_store_download` or
+        `/api/histories/{id}/write_store` instead.
```
Comment (jmchilton) [inline lib/galaxy/webapps/galaxy/api/histories.py:429]: ❤️

File: lib/galaxy/webapps/galaxy/api/tasks.py [hunk 4]
```diff
@@ -14,11 +17,12 @@
 
 @router.cbv
 class FastAPITasks:
+    manager: AsyncTasksManager = depends(AsyncTasksManager)  # type: ignore[type-abstract]
```
Comment (jmchilton) [inline lib/galaxy/webapps/galaxy/api/tasks.py:20]: Over and over mypy just sucks on this point - I want to ``--disable-error-code=type-abstract`` this whole project. xref https://github.com/python/mypy/issues/4717#issuecomment-1264388209. I don't think I've ever seen a bug found via that issue - it just makes good code and code patterns look ugly.

File: lib/galaxy/app.py [hunk 5]
```diff
@@ -538,6 +542,10 @@ def __init__(self, configure_logging=True, use_converters=True, use_display_appl
         self.role_manager = self._register_singleton(RoleManager)
         self.job_manager = self._register_singleton(JobManager)
 
+        self.task_manager = self._register_abstract_singleton(
```
Comment (jmchilton) [inline lib/galaxy/app.py:545]: Wonderful! I love this!

**Comment by jmchilton** [comment]
This is utterly wonderful - you're amazing!

File: lib/galaxy_test/selenium/test_history_export.py (hunk 2 above)
Comment (davelopez) [reply lib/galaxy_test/selenium/test_history_export.py:12]: > Is this definitely being used?

Yes, it does work, see the call stack in the screenshot to find where is being called:
image

But I see your point... this kind of turned into an integration test with this change... should I just move it to test/integration_selenium then?

File: lib/galaxy/webapps/galaxy/api/tasks.py (hunk 4 above)
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/api/tasks.py:20]: Yeah, it's a bit disappointing that mypy doesn't handle this nicely...

**Comment by davelopez** [comment]
Thank you so much @jmchilton, you are the amazing one! :D

File: lib/galaxy_test/selenium/test_history_export.py (hunk 2 above)
Comment (davelopez) [reply lib/galaxy_test/selenium/test_history_export.py:12]: Since Celery is enabled by default in the test instance I need it to be disabled to keep testing the previous UI, so I decided to move it and convert it into an integration test for consistency.

Thank you John! let me know if I should approach it differently.

**Comment by davelopez** [comment]
This is finally green! :)

Let me know if there is something I can further improve or fix.
I will follow up with some reactivity improvements once https://github.com/galaxyproject/galaxy/pull/15108 is ready.

**Review by mvdbeek — APPROVED** [review]
Absolutely amazing work, thanks so much @davelopez!
